### PR TITLE
Simplify release publishing

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR adds a `publish-release` target to the `Makefile` that uses [Goreleaser](https://goreleaser.com/) to simplify the publishing of new releases.
This was used to generate [v0.5.0](https://github.com/mattermost/mattermost-mattermod/releases/tag/v0.5.0).

### CODE REVIEW

**Review by agnivade — COMMENTED** [review]
Can we also add the publish-release target after the patch, minor and major targets so that a release gets generated automatically once a tag is created? Or does it need the tag to be pushed first to GitHub? In which case, I'd prefer some `@echo` commands nudging the operator to run publish-release.

**Comment by streamer45** [comment]
> Can we also add the publish-release target after the patch, minor and major targets so that a release gets generated automatically once a tag is created? Or does it need the tag to be pushed first to GitHub? In which case, I'd prefer some `@echo` commands nudging the operator to run publish-release.

The tag just has to exist but I see those targets do commit and push it so running `publish-release` after should work just fine.

**Review by agnivade — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -164,6 +164,21 @@ minor: release
...
+		echo "goreleaser is not installed, do you want to download it? [y/N] " && read ans && [ $${ans:-N} = y ]; \
+		if [ $$ans = y ] || [ $$ans = Y ]  ; then \
+			curl -sfL https://install.goreleaser.com/github.com/goreleaser/goreleaser.sh | sh; \
+		else \
+			echo "aborting make publish-release."; \
+			exit -1; \
+		fi; \
+	else \
+		goreleaser --rm-dist; \
+	fi; \
```
Comment (isacikgoz) [inline Makefile:180]: We need to remove this `else` block, If I don't have `goreleaser` and install it here, I need to run `make publish-release` again to make this work.

Comment (streamer45) [reply Makefile:180]: Oh I see. It was meant to install and run at the same time, I'll move it after the `curl | sh` bit then.

Comment (streamer45) [reply Makefile:180]: One a second though. Might as well keep it at the end since we exit on failure.

Comment (isacikgoz) [reply Makefile]: Yeah, definitely :)

**Review by isacikgoz — APPROVED** [review]
Awesome! 🚀

File: .goreleaser.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+before:
+  hooks:
+    - go mod download
+    - go mod verify
+builds:
+  - main: ./cmd/mattermost-mattermod
```
Comment (khos2ow) [inline .goreleaser.yml:6]: We don't necessarily need to release this binary since we already publish docker image.

File: .goreleaser.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    flags:
+      - -mod=readonly
+      - -trimpath
+    ldflags:
+      - -X github.com/mattermost/mattermost-mattermod/version.version=v{{.Version}}
+      - -X github.com/mattermost/mattermost-mattermod/version.commitHash={{.FullCommit}}
+      - -X github.com/mattermost/mattermost-mattermod/version.buildDate={{.Date}}
+    env:
+      - CGO_ENABLED=0
+  - main: ./cmd/jobserver
```
Comment (khos2ow) [inline .goreleaser.yml:22]: The same here.

File: .goreleaser.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      - -X github.com/mattermost/mattermost-mattermod/version.version=v{{.Version}}
+      - -X github.com/mattermost/mattermost-mattermod/version.commitHash={{.FullCommit}}
+      - -X github.com/mattermost/mattermost-mattermod/version.buildDate={{.Date}}
+    env:
+      - CGO_ENABLED=0
+  - main: ./cmd/migrator
+    id: "migrator"
+    binary: bin/migrator
+    goos:
+      - linux
```
Comment (khos2ow) [inline .goreleaser.yml:42]: This is essentially what's needed and it's a tool for operators so we need to publish all available os (`darwin`, `windows`, `linux` and maybe just throwing `freebsd` out there too.)

File: .goreleaser.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    goarch:
+      - amd64
```
Comment (khos2ow) [inline .goreleaser.yml:44]: It's nice to have `arm` too.

File: .goreleaser.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      - CGO_ENABLED=0
+archives:
+  -
+    name_template: "{{ .ProjectName }}-v{{ .Version }}-{{ .Os }}-{{ .Arch }}"
+    wrap_in_directory: true
+    format: tar.gz
+    files:
+      - LICENSE
+      - README.md
+      - config/config-mattermod.default.json
```
Comment (khos2ow) [inline .goreleaser.yml:62]: We shouldn't bundle config with the release artifact, it gives an impression that that config is needed (which is not, it's just an in-tree sample)

File: Makefile [hunk 7]
```diff
@@ -152,17 +152,31 @@ release:
...
+	@if ! [ -x "$$(command -v goreleaser)" ]; then \
+		echo "goreleaser is not installed, do you want to download it? [y/N] " && read ans && [ $${ans:-N} = y ]; \
+		if [ $$ans = y ] || [ $$ans = Y ]  ; then \
+			curl -sfL https://install.goreleaser.com/github.com/goreleaser/goreleaser.sh | sh; \
+		else \
+			echo "aborting make publish-release."; \
+			exit -1; \
+		fi; \
+	fi; \
+	goreleaser --rm-dist; \
```
Comment (khos2ow) [inline Makefile:179]: Leftover `\` at the end of line.

**Review by khos2ow — COMMENTED** [review]
It's already merged but I have some comments below.

**Pull request merged**